sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala

```diff
@@ -66,55 +67,23 @@ case class ScriptTransformation(
       val inputStream = proc.getInputStream
       val outputStream = proc.getOutputStream
       val errorStream = proc.getErrorStream
-      val reader = new BufferedReader(new InputStreamReader(inputStream))
-
       val (outputSerde, outputSoi) = ioschema.initOutputSerDe(output)
 
       val iterator: Iterator[InternalRow] = new Iterator[InternalRow] with HiveInspectors {
-        var cacheRow: InternalRow = null
-        var curLine: String = null
-        var eof: Boolean = false
-
+        val scriptOutputStream = new DataInputStream(inputStream)
+        var scriptOutputWritable: Writable = null
+        // This method blocks until the inputdata availble or reach EOF
         override def hasNext: Boolean = {
-          if (outputSerde == null) {
-            if (curLine == null) {
-              curLine = reader.readLine()
-              curLine != null
-            } else {
+          if (scriptOutputWritable == null) {
+            scriptOutputWritable = outputSerde.getSerializedClass().newInstance
```

Comment:
We don't need to create the `scriptOutputWritable` instance for every input row; we can create it outer of the `hasNext`.